components/dash-core-components/src/fragments/Graph.react.js

```diff
@@ -227,7 +228,39 @@ class PlotlyGraph extends Component {
     }
 
     getConfig(config, responsive) {
-        return mergeDeepRight(config, this.getConfigOverride(responsive));
+        const modeBarButtons = this.getModeBarButtons(config);
+        if (modeBarButtons) {
+            this.configModeBarButtons = modeBarButtons;
+        }
+
+        let configClone = mergeDeepRight(
+            config,
+            this.getConfigOverride(responsive)
+        );
+
+        if (this.configModeBarButtons) {
+            configClone = mergeDeepRight(configClone, {
+                modeBarButtons: this.configModeBarButtons,
```

Comment:
Is there a reason for the `this.configModeBarButtons` indirection? Looks like the idea might have been that from one update to the next if you removed `modeBarButtons` from the `config` (or removed `config` entirely?) the previous set of buttons would remain? I don't see why we'd want that, but also this wouldn't actually happen because `this.getModeBarButtons` will always return a truthy value - unlike in Python, in JS `[]` is truthy. Likewise we're always going to enter the `if (this.configModeBarButtons)` branch here.

So I'd suggest to remove `this.configModeBarButtons` entirely, and have this branch be `if (modeBarButtons.length)`

Reply:
This was done because the configuration is overwritten inside the component. If we have an empty parameter `modeBarButtons` then it remains empty and component get default config. But if there is a string configuration there, then after initialization  the string is replaced with a JS object and the click function is not passed to the python code. So I made local storage for the string parameter so that it can be reused.

Reply:
> the configuration is overwritten inside the component

That's why we're cloning the config and copying in the original `modeBarButtons` spec, but I still don't see the purpose of `this.configModeBarButtons`. Can you show a situation (even better if it's a test!) where we need this bit of internal state to get the right behavior? I'm just wary of it holding on to a state we no longer want and thus introducing a new more subtle bug.

Reply:
You can check example code from issue #1157. The problem manifests itself when trying to copy or move a component. It turns out that such parameters `modeBarButtons=[['resetScale2d']]` are transferred during initilization, and such `'modeBarButtons': [[{'name': 'resetScale2d', '_cat': 'resetscale', 'attr': 'zoom', 'val': 'reset', 'icon': {'width': 928.6, 'height': 1000, 'path': 'm786 296v-267q0-15-11-26t-25-10h-214v214h-143v-214h-214q-15 0-25 10t-11 26v267q0 1 0 2t0 2l321 264 321-264q1-1 1-4z m124 39l-34-41q-5-5-12-6h-2q-7 0-12 3l-386 322-386-322q-7-4-13-4-7 2-12 7l-35 41q-4 5-3 13t6 12l401 334q18 15 42 15t43-15l136-114v109q0 8 5 13t13 5h107q8 0 13-5t5-13v-227l122-102q5-5 6-12t-4-13z', 'transform': 'matrix(1 0 0 -1 0 850)'}}]]},` are returned when copying the configuration.